---
theme: post
date:  01-02-2021
title: "JS: Asynchronous"
---

# Async Javascript

JS synchronicity took me a while at the beginning to wrap my head around, so in the spirit of this blog  
let's have a look at it

## Old School Style 

It is worth noting that in the old school way, this used to be handled with callbacks. 

Maybe at some point I will touch on how these used to be done but for now, let's simply
look at the modern method. 

## Promises

`Promises` are a neat way of resolving asynchronous code. A `Promise` is a function
that takes two functions as arguments, and has some special properties. The two functions
taken as arguments are `(resolve, reject)` and the arguments that are given to these 
are the arguments returned by the `Promise`

Let's look at an example:

```javascript
const promise = new Promise((resolve, reject) => {
    const a = 1 + 3;
    if (a == 4){
        resolve(a);
    } else {
        reject('Math does not work anymore');
    }
})
```
This Promise will resolve to `a` which in this case is `4`. The way we can access the 
result of the promise is by the use of `then` and `catch`. Like so

```javascript
promise.then(message => console.log("Resolve: ", message))
       .catch(message => console.log ("Reject: ",message)) // this would print if the Promise calls reject()
```

And believe it or not that is it. 

### Why though?

JS is single threaded. That means it executes code in the order it is written. By calling a `Promise`, the code 
inside the Promise then is executed in a separate thread, allowing it to be ran in parallel. 

---

## Async functions

`async` functions are an easier way of returning a promise. Let's have a quick look:

```javascript
const printVlad = () => new Promise((resolve,err) => {resolve("Vlad"); err("Error");})
```

The above function `printVlad` creates a `Promise` that resolves to `"Vlad"`. But the
async notation allows us to take this and write it in a simpler manner 

```javascript
const printVlad = async () => "Vlad"
```

Yes, that is a lot more compact. But what if we wanted to be able to specify what happens if the 
promise fails. Well for this you can always explicitly specify the `Promise.resolve` and `Promise.reject`.

```javascript
const printVlad = async () => { Promise.resolve("Vlad"); Promise.reject("Error!")};
```
